### PR TITLE
Log Levels View

### PR DESCRIPTION
**Pull request by vudayani-vmw**

### CODE REVIEW

File: vscode-extensions/vscode-spring-boot/lib/set-log-levels-ui.ts [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+/** Called when extension is activated */
+export function activate(
+        client: LanguageClient,
+        options: ActivatorOptions,
+        context: VSCode.ExtensionContext
+) {
+    context.subscriptions.push(
+        VSCode.commands.registerCommand('vscode-spring-boot.set.log-levels', () => {
+            if (client.isRunning()) {
+                client.onNotification(LiveProcessLoggersUpdatedNotification.type, getLoggersList)
```
Comment (BoykoAlex) [inline vscode-extensions/vscode-spring-boot/lib/set-log-levels-ui.ts:95]: Hmm... is this inside the command handler? Might not be a good idea to register notification listeners every time command is executed...

**Review by BoykoAlex — COMMENTED** [review]
Some initial review... don't think I follow everything in there... I think i was most confused by `getLoggers` and `getLoggersData` thing...

File: headless-services/spring-boot-language-server/src/main/java/org/springframework/ide/vscode/boot/java/livehover/v2/SpringProcessCommandHandler.java [hunk 2]
```diff
@@ -43,6 +44,9 @@ public class SpringProcessCommandHandler {
 	private static final String COMMAND_LIST_CONNECTED = "sts/livedata/listConnected";
 	private static final String COMMAND_GET_METRICS = "sts/livedata/get/metrics";
 	private static final String COMMAND_GET_REFRESH_METRICS = "sts/livedata/refresh/metrics";
+	private static final String COMMAND_GET_LOGGERS = "sts/livedata/getLoggers";
+	private static final String COMMAND_FETCH_LOGGERS_DATA = "sts/livedata/fetch/loggersData";
```
[the remark this answers is not in the document]
Comment (vudayani-vmw) [reply headless-services/spring-boot-language-server/src/main/java/org/springframework/ide/vscode/boot/java/livehover/v2/SpringProcessCommandHandler.java:48]: The `COMMAND_GET_LOGGERS = "sts/livedata/getLoggers"` - is to make a get request or JMX Bean operation to get the list of loggers async. Once the loggers are fetched it is stored in a [loggersData map](https://github.com/spring-projects/sts4/blob/a2802d1e39b2fff4408db4cf5440e9672f8b6928/headless-services/spring-boot-language-server/src/main/java/org/springframework/ide/vscode/boot/java/livehover/v2/SpringProcessLiveDataProvider.java#L81) and a notification is sent to the client that logger data is available. 
The `COMMAND_FETCH_LOGGERS_DATA = "sts/livedata/fetch/loggersData"` - is used to fetch this data from the map and show it in the IDE
The naming convention above is bad i guess and it is causing the confusion.

File: headless-services/commons/commons-lsp-extensions/src/main/java/org/springframework/ide/vscode/commons/protocol/STS4LanguageClient.java [hunk 3]
```diff
@@ -51,6 +51,12 @@ public interface STS4LanguageClient extends LanguageClient, SpringIndexLanguageC
 	
 	@JsonNotification("sts/liveprocess/gcpauses/metrics/updated")
 	void liveProcessGcPausesMetricsDataUpdated(LiveProcessSummary processKey);
+	
+	@JsonNotification("sts/liveprocess/loggers/updated")
```
[the remark this answers is not in the document]
Comment (vudayani-vmw) [reply headless-services/commons/commons-lsp-extensions/src/main/java/org/springframework/ide/vscode/commons/protocol/STS4LanguageClient.java:55]: > What are these 2 methods are doing? My guess is once action is executed to refresh running process live data notification is sent out to the client to refresh UI etc? I see similar method for metrics but forgot the purpose :-\ I might figure it out while reviewing the rest of the code though ;-)

Yes, that's correct. It is to send notification to the client that actuator data is available. Fetch and refresh the UI.

> Does the notification with LiveProcessLoggersSummary parameter simply lets IDE know that the log level that was attempted to be changed with a post request or JMX Bean operation has indeed changed? Is that it?

Yes, it just contains additional fields to let IDE know which package and log level was updated. This information is displayed as a notification in the IDE. 
I didn't want to store this data and make another call to fetch the changed log level info to display in the UI as a notification.

File: vscode-extensions/vscode-spring-boot/lib/set-log-levels-ui.ts [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+                effectiveLevel,
+                label
+            };
+        });
+    }
+    if(items) {
+        const chosenPackage = await VSCode.window.showQuickPick(items);
+        if (chosenPackage) {
+            const chosenlogLevel = await VSCode.window.showQuickPick(loggers.levels);
+            const changeLogLevel = await VSCode.commands.executeCommand('sts/livedata/configure/logLevel', {"endpoint": "loggers"}, process, chosenPackage, {"configuredLevel":chosenlogLevel});
```
Comment (BoykoAlex) [inline vscode-extensions/vscode-spring-boot/lib/set-log-levels-ui.ts:75]: Feels like `{"endpoint": "loggers"}` is a redundant piece of info.

**Review by BoykoAlex — COMMENTED** [review]
Generally all looks good to me. I only had one comment about the code. The tests started passign for me once i rebased on the latest from **main** branch. I have tried the UI in the vscode - it worked just fine. Thus, overall, it all looks good.

There is however a broader question i had. My understanding is that we cache the loggers once receive them via JMX/Actuator endpoint and send a notification that data is there. I was thinking that perhaps unlike with memory metrics and other things we don't need to cache the loggers data? If the plan is just list current loggers in the UI for the set logger level action then we could just have a fetch command to show them within the action UI and then forget. While loggers are fetched indefinite progress bar is shown in the show pick UI with some message that loggers are fetched etc. What do you think?

File: headless-services/spring-boot-language-server/src/main/java/org/springframework/ide/vscode/boot/java/livehover/v2/SpringProcessConnectorService.java [hunk 5]
```diff
@@ -292,4 +297,130 @@ private void scheduleRefresh(IndefiniteProgressTask progressTask, SpringProcessP
...
 	}
+	
+	public void getLoggers(SpringProcessParams springProcessParams) {
+		log.info("get loggers data: " + springProcessParams.getProcessKey());
+
+		SpringProcessConnector connector = this.connectors.get(springProcessParams.getProcessKey());
+		if (connector != null) {
+			final IndefiniteProgressTask progressTask = getProgressTask(
+					"spring-process-connector-service-get-loggers-data" + springProcessParams.getProcessKey(), "Loggers", null);
+			System.out.println(progressTask);
```
Comment (BoykoAlex) [inline headless-services/spring-boot-language-server/src/main/java/org/springframework/ide/vscode/boot/java/livehover/v2/SpringProcessConnectorService.java:308]: Watch out for these `System.out.println` statements in the code :-) They are okay when you connect to LS via socket but once you build a VSCode extension and launch it without connecting to LS the comm goes over System in/out and that print statement would just corrupt the JSON RPC comm and shutdown the LS for unknown reason ;-)
So, yes, these statement would need to go or replaced with log statement :-)
I'm guilty for leaving `System.out.println` multiple times ;-)

File: vscode-extensions/vscode-spring-boot/lib/set-log-levels-ui.ts (hunk 4 above)
Comment (vudayani-vmw) [reply vscode-extensions/vscode-spring-boot/lib/set-log-levels-ui.ts]: Updated getLoggers and removed caching logic. A progress bar is shown in the status bar of the window when loggers are being fetched.

**Comment by BoykoAlex** [comment]
@vudayani-vmw Just tried it today. Seems to be working and looks good overall. I'd just add the copyright headers for new files, mark the PR as "ready" and merge it.

**Comment by martinlippert** [comment]
@vudayani-vmw @BoykoAlex Keep in mind that newly added copyright headers should refer to `Broadcom, Inc.` instead of `VMware, Inc.`

**Comment by BoykoAlex** [comment]
LGTM - squashing and merging. Thanks @vudayani-vmw - great job!

**Pull request merged**